dali/pipeline/operator/sequence_operator.h

```diff
@@ -0,0 +1,627 @@
+    SetupSequenceOperator(ws);
+    is_expanding_ = ShouldExpand(ws);
+    bool is_inferred;
+    if (!IsExpanding()) {
+      is_inferred = Operator<Backend>::Setup(output_desc, ws);
+    } else {
+      DALI_ENFORCE(IsExpandable(),
+                   "Operator requested to expand the sequence-like inputs, but no expandable input "
+                   "was found");
+      SetupExpandedWorkspace(expanded_, ws);
+      ExpandInputs(ws);
+      ExpandArguments(ws);
+      is_inferred = Operator<Backend>::Setup(output_desc, expanded_);
+    }
+    is_inferred_ = ProcessOutputDesc(output_desc, ws, is_inferred);
+    return is_inferred_;
+  }
+
+  void Run(workspace_t<Backend> &ws) override {
+    if (!IsExpanding()) {
+      Operator<Backend>::Run(ws);
+    } else {
+      if (is_inferred_) {
+        ExpandOutputs(ws);
+      } else {
+        // nothing could be presized, so there is nothing to expand
+        // just copy the handles from ``ws`` to ``expanded_``
+        ShareOutputs(ws);
+      }
+      Operator<Backend>::Run(expanded_);
+      if (is_inferred_) {
+        PostprocessOutputs(ws);
+      }
+      ClearExpanded();
+    }
+  }
+
+  bool IsExpanding() const {
+    return is_expanding_;
+  }
+
+  bool IsExpandable() const {
+    auto expand_like_idx = ExpandLikeIdx();
+    assert(expand_like_idx == -1 || GetInputExpandDesc(expand_like_idx).NumDimsToExpand() > 0);
+    return expand_like_idx >= 0;
+  }
+
+ protected:
+  /**
+   * @brief Controls if given input should expand leading channel dimensions.
+   *
+   * If false, only a batch whose layout starts with ``F`` will be considered expandable
+   * (for example ``FHWC -> HWC``). If true, layouts starting with ``C`` will be expandable as well,
+   * i.e. ``FCHW -> HW, CFHW -> HW, CHW -> HW`` etc.
+   */
+  virtual bool ShouldExpandChannels(int input_idx) const {
+    (void)input_idx;
+    return false;
+  }
+
+  /**
+   * @brief Controls if the SequenceOperator should expand batches in the current iteration.
+   *
+   * It can be used to restrict cases when the expansion is performed, but it is an error for
+   * ``ShouldExpand`` to return true if none of the inputs is expandable (i.e. !IsExpandable()).
+   */
+  virtual bool ShouldExpand(const workspace_t<Backend> &ws) {
+    return IsExpandable();
+  }
+
+  /**
+   * @brief Expands inputs from the provided workspace and adds them to the expanded workspace.
+   *
+   * Assumes ``IsExpandble()`` is true, i.e. there is at least one expandable input.
+   *
+   * For multiple-input operators, if any of the inputs is to be expanded, all other inputs
+   * must be expandable in an agreeable way, i.e:
+   * 1. have the same expandble layout prefix and matching outermost dimensions that are to
+   *    be expanded, or
+   * 2. have no expandable layout prefix, in that case the samples are going to be broadcasted
+   *    to match the other expanded inputs (TODO).
+   */
+  virtual void ExpandInputs(const workspace_t<Backend> &ws) {
```

Comment:
I moved it a step down the ladder to (nonvirtual)ExpandInputs - calls in a loop -> virtual ExpandInput(). Some basic validation was moved to the Setup method, the rest - specific to the way expansion is done - to methods called by ExpandInput, so that it may be reused only if applicable.